test/language/expressions/class/private-methods/prod-private-async-generator.js

```diff
@@ -0,0 +1,142 @@
+
+    ---
+
+    InitializeClassElements ( F, proto )
+
+    ...
+    5. For each item element in order from elements,
+      a. Assert: If element.[[Placement]] is "prototype" or "static", then element.[[Key]] is not a Private Name.
+      b. If element.[[Kind]] is "method" and element.[[Placement]] is "static" or "prototype",
+        i. Let receiver be F if element.[[Placement]] is "static", else let receiver be proto.
+        ii. Perform ? DefineClassElement(receiver, element).
+
+    InitializeInstanceElements ( O, constructor )
+
+    ...
+    3. Let elements be the value of F's [[Elements]] internal slot.
+    4. For each item element in order from elements,
+      a. If element.[[Placement]] is "own" and element.[[Kind]] is "method",
+        i. Perform ? DefineClassElement(O, element).
+
+    DefineClassElement (receiver, element)
+
+    ...
+    6. If key is a Private Name,
+      a. Perform ? PrivateFieldDefine(receiver, key, descriptor).
+
+    PrivateFieldDefine (P, O, desc)
+
+    ...
+    6. Append { [[PrivateName]]: P, [[PrivateFieldDescriptor]]: desc } to O.[[PrivateFieldDescriptors]].
+
+---*/
+var ctorPromise;
+
+
+
+/***
+ * template notes:
+ * 1. method should always be #m
+ * 2. the template provides c.ref/other.ref for external reference
+ */
+
+function hasOwnProperty(obj, name) {
+  return Object.prototype.hasOwnProperty.call(obj, name);
+}
+
+var C = class {
+  async * #m() { return 42; }
+
+
+  get ref() { return this.#m; }
+
+  constructor() {
+    assert.sameValue(
+      hasOwnProperty(this, '#m'), false,
+      'private methods are defined in an special internal slot and cannot be found as own properties'
+    );
+    assert.sameValue(typeof this.#m, 'function');
+    assert.sameValue(this.ref(), this.#m, 'returns the same value');
+
+    var ctorIter = this.#m();
+    var p = ctorIter.next();
+    ctorPromise = p.then(({ value, done }) => {
+        assert.sameValue(value, 42, 'return from generator method, inside ctor');
+        assert.sameValue(done, true, 'iterator is done, inside ctor');
+    }, $DONE);
+    assert.sameValue(this.#m.name, '#m', 'function name inside constructor');
+
+  }
+}
+
+var c = new C();
+var other = new C();
+
+assert.sameValue(
+  hasOwnProperty(C.prototype, '#m'), false,
+  'method is not defined in the prototype'
+);
+
+assert.sameValue(
+  hasOwnProperty(C, '#m'), false,
+  'method is not defined in the contructor'
+);
+
+assert.sameValue(
+  hasOwnProperty(c, '#m'), false,
+  'method cannot be seen outside of the class'
+);
+
+/***
+ * MethodDefinition : ClassElementName ( UniqueFormalParameters ) { FunctionBody }
+ * 
+ * 1. Let methodDef be DefineMethod of MethodDefinition with argument homeObject.
+ * ...
+ */
+assert.sameValue(c.ref, other.ref, 'The method is defined once, and reused on every new instance');
+
+assert.sameValue(c.ref.name, '#m', 'function name is preserved external reference');
+ctorPromise.then(() => {
+    var iter = c.ref();
```

Comment:
I think it would be helpful if there was a comment that accompanied this line, which noted that this is actually calling `this.#m`